File: stage3_temp_quantiles.py
#!/usr/bin/env python3
#
# Generate percentiles 0-100 of tmax and tmin files. Low-memory verison.
#
# Contact: Edgar Castro <edgar_castro@g.harvard.edu>

import csv
import collections
import os
import gzip
import typing

import numpy
import tqdm

DEFAULT_TMAX_FILENAME = "mean_tmax.csv.gz"
DEFAULT_TMIN_FILENAME = "mean_tmin.csv.gz"
DEFAULT_TMAX_QUANTILES_FILENAME = "tmax_quantiles.csv.gz"
DEFAULT_TMIN_QUANTILES_FILENAME = "tmin_quantiles.csv.gz"

DEFAULT_PERCENTILES = range(101)

LAST_YEAR = 2020  # For updating the progress bar


def dump_binned_values(values_by_id,
                       current_year,
                       writer,
                       percentiles: typing.Iterable[int] = DEFAULT_PERCENTILES):
    for id_, values in tqdm.tqdm(
            values_by_id.items(),
            position=1,
            desc="Generating pctiles for {}".format(current_year)
    ):
        writer.writerow(
            [id_, current_year] + list(numpy.percentile(
                values, percentiles
            ))
        )


def extract_quantiles(input_path: str, output_path: str):
    with gzip.open(input_path, "rt") as input_fp:
        reader = csv.DictReader(input_fp)
        id_column = reader.fieldnames[0]

        current_year = None
        values_by_id = collections.defaultdict(list)

        temp_path = "{}.part".format(output_path)

        with gzip.open(temp_path, "wt") as output_fp:
            writer = csv.writer(output_fp)
            writer.writerow([id_column, "year"] + ["pctile{:02d}".format(i) for i in range(101)])

            iterator = tqdm.tqdm(
                reader,
                position=0,
                desc="Reading data",
                smoothing=0
            )

            for row in iterator:
                year = row["date"][:4]

                if year != current_year:
                    if current_year is not None:

                        dump_binned_values(values_by_id, current_year, writer)

                        # If not initialized: update the iterator so we can have ETA
                        if iterator.total is None:
                            total_rows = (
                                    len(values_by_id)
                                    * (LAST_YEAR - 1980 + 1)
                                    * 365
                            )
                            tqdm.tqdm.write("Estimated {} total rows".format(total_rows))
                            iterator.total = total_rows

                    # Reset bins
                    values_by_id = collections.defaultdict(list)
                    current_year = year

                values_by_id[row[id_column]].append(float(row["value"]))

            # Dump last year
            dump_binned_values(values_by_id, current_year, writer)

        os.rename(temp_path, output_path)

if __name__ == "__main__":
    import argparse
    import glob

    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input", default=None)
    parser.add_argument("-o", "--output", default=None)
    args = parser.parse_args()

    # Input given: convert the given file
    if all([args.input, args.output]):
        extract_quantiles(args.input, args.output)

    # No input given: determine what files need to be converted by looking for
    # missing files in the expected daymet-aggregation output directory
    # hierarchy
    else:
        for aggregated_directory in glob.glob("output/aggregated-combined/*"):
            extra_directory = aggregated_directory.replace("aggregated-combined", "extra")

            tmax_path = os.path.join(aggregated_directory, DEFAULT_TMAX_FILENAME)
            tmin_path = os.path.join(aggregated_directory, DEFAULT_TMIN_FILENAME)
            tmax_quantiles_path = os.path.join(extra_directory, DEFAULT_TMAX_QUANTILES_FILENAME)
            tmin_quantiles_path = os.path.join(extra_directory, DEFAULT_TMIN_QUANTILES_FILENAME)

            for path in [tmax_path, tmin_path]:
                if not os.path.isfile(path):
                    raise Exception("ERROR: {} does not exist".format(path))

            if not os.path.isdir(extra_directory):
                os.makedirs(extra_directory)

            if os.path.isfile(tmax_quantiles_path):
                print("Skipping {}".format(tmax_quantiles_path))
            else:
                print("Generating {}".format(tmax_quantiles_path))
                extract_quantiles(tmax_path, tmax_quantiles_path)

            if os.path.isfile(tmin_quantiles_path):
                print("Skipping {}".format(tmin_quantiles_path))
            else:
                print("Generating {}".format(tmin_quantiles_path))
                extract_quantiles(tmin_path, tmin_quantiles_path)

File: stage3_temp_quantiles.R
#!/usr/bin/env Rscript
#
# Generate percentiles 0-100 of tmax and tmin files.
#
# Contact: Edgar Castro <edgar_castro@g.harvard.edu>

library(data.table)
library(pbapply)
library(unglue)

pboptions(type = "timer")

paths <- Sys.glob("output/aggregated-combined/*/*.csv.gz")
parts <- as.data.table(unglue_data(
  paths,
  "output/aggregated-combined/{geography}/{aggregation}_{measure}.csv.gz"
))[, path := paths][aggregation == "mean"]

parts <- parts[!grepl("block_group", path)]

PERCENTILES <- 0:100 / 100

for (current_geography in unique(parts$geography)) {
  output_directory <- sprintf("output/extra/%s", current_geography)
  dir.create(output_directory, showWarnings = FALSE, recursive = TRUE)
  
  output_tmax_quantiles <- sprintf("%s/tmax_quantiles.csv.gz", output_directory)
  output_tmin_quantiles <- sprintf("%s/tmin_quantiles.csv.gz", output_directory)
  
  message(sprintf("Processing: %s", current_geography))
  
  if (file.exists(output_tmax_quantiles)) {
    message("(1/8) Already done")
    message("(2/8) Already done")
    message("(3/8) Already done")
    message("(4/8) Already done")
  } else {
    message("(1/8) Reading and transforming tmax data")
    tmax <- fread(parts[geography == current_geography & measure == "tmax"]$path)
    
    # Hacky: do some math to extract the year, month, and day from ISO8601 dates
    # with no delimiter
    tmax[, year := floor(date / 1e4)]
    tmax[, month := floor((date / 1e4 - year) * 1e2)]
    tmax[, day := date - (floor(date / 1e2) * 1e2)]
    
    # Guess the ID column (should be the first column) and rename it (syntactically
    # inconvenient to use variables in data.table `by` argument)
    id_column <- names(tmax)[1]
    setnames(tmax, id_column, "id")
    
    # Generate quantiles ----
    message("(2/8) Calculating tmax percentiles")
    tmax_agg <- pblapply(
      PERCENTILES,
      function(percentile) {
        result <- tmax[, list(temp = quantile(value, percentile, na.rm = TRUE)),
                       by = list(id, year)]
        setnames(result, "temp", sprintf("pctile%02.f", percentile * 100))
      }
    )
    
    # Merge and write out ----
    message("(3/8) Merging tmax percentiles")
    tmax_agg <- Reduce(function(left, right) cbind(left, right[, 3]), tmax_agg)
    setnames(tmax_agg, "id", id_column)
    
    message("(4/8) Writing tmax percentiles")
    fwrite(tmax_agg, output_tmax_quantiles)
    
    rm(tmax)
    rm(tmax_agg)
    gc()
  }
  
  # Repeat for tmin
  if (file.exists(output_tmin_quantiles)) {
    message("(5/8) Already done")
    message("(6/8) Already done")
    message("(7/8) Already done")
    message("(8/8) Already done")
  } else {
    message("(5/8) Reading and transforming tmin data")
    tmin <- fread(parts[geography == current_geography & measure == "tmin"]$path)
    
    # Hacky: do some math to extract the year, month, and day from ISO8601 dates
    # with no delimiter
    tmin[, year := floor(date / 1e4)]
    tmin[, month := floor((date / 1e4 - year) * 1e2)]
    tmin[, day := date - (floor(date / 1e2) * 1e2)]
    
    # Guess the ID column (should be the first column) and rename it (syntactically
    # inconvenient to use variables in data.table `by` argument)
    id_column <- names(tmin)[1]
    setnames(tmin, id_column, "id")
    
    # Generate quantiles ----
    message("(6/8) Calculating tmin percentiles")
    tmin_agg <- pblapply(
      PERCENTILES,
      function(percentile) {
        result <- tmin[, list(temp = quantile(value, percentile, na.rm = TRUE)),
                       by = list(id, year)]
        setnames(result, "temp", sprintf("pctile%02.f", percentile * 100))
      }
    )
    
    # Merge and write out ----
    message("(7/8) Merging tmin percentiles")
    tmin_agg <- Reduce(function(left, right) cbind(left, right[, 3]), tmin_agg)
    setnames(tmin_agg, "id", id_column)
    
    message("(8/8) Writing tmin percentiles")
    fwrite(tmin_agg, output_tmin_quantiles)
    
    rm(tmin)
    rm(tmin_agg)
    gc()
  }
}
